### PR TITLE
feature: add option to retain .vcs

### LINKED ISSUE #8746: packages using git information during build fail when cached
During caching, the `.git` directory is excluded, effectively archiving the package without using git archive for some reason.  The result is that packages that depend on git information to determine their version at configure time, like flux and possibly rose as in #3749, break when cached but not when checked out initially.

### Expected Result

That the cached version will be the same as the non-cached version.

### Actual Result

The `.git` directory is removed.

### PR DESCRIPTION
**Pull request by matz-e**
Some packages require version control information to build, e.g., python
packages depending on `setuptools-scm`.  Spack will remove any VCS
directory when caching hashable stages like tags, this patch will
provide a parameter `clean=False` to the `version` function that will
retain `.git` etc.

Should fix #8746.

### CODE REVIEW

**Comment by ReinhardPrix** [comment]
Hi, so what's the status for this PR? I would urgently need that feature as well, as I'm struggling to create a package that has exactly that issue (ie build fails with nonexistant .git subdir).

**Comment by matz-e** [comment]
> Hi, so what's the status for this PR? I would urgently need that feature as well, as I'm struggling to create a package that has exactly that issue (ie build fails with nonexistant .git subdir).

We're using this feature very reliably with our own code base/central deployment. Not sure who to tag for review here. Maybe @alalazo?

**Comment by ReinhardPrix** [comment]
> We're using this feature very reliably with our own code base/central deployment. Not sure who to tag for review here. Maybe @alalazo?

Thanks. I can confirm that your feature branch also works fine for me, but would be nice to get this merged into main as well.

**Comment by ReinhardPrix** [comment]
Maybe one could add a unit-test using that new feature, in order to get the code-coverage failing test to pass? Not sure what the merge-policy here is, but maybe that's why it's stuck right now?

File: lib/spack/spack/fetch_strategy.py [hunk 1]
```diff
@@ -502,7 +503,7 @@ def archive(self, destination, **kwargs):
         tar = which('tar', required=True)
 
         patterns = kwargs.get('exclude', None)
```
Comment (alalazo) [inline lib/spack/spack/fetch_strategy.py]: Looking at the code in this function it seems that we can exclude patterns that are unrelated to VCS. In that sense what is written in the docs might be inexact, as one could get those patterns into the archive by giving `clean=False`. Can we target VCS patterns in a more specific way?

File: lib/spack/spack/fetch_strategy.py [hunk 2]
```diff
@@ -478,6 +478,7 @@ def __init__(self, **kwargs):
 
         # Set a URL based on the type of fetch strategy.
         self.url = kwargs.get(self.url_attr, None)
+        self.clean = kwargs.get("clean", True)
```
Comment (alalazo) [inline lib/spack/spack/fetch_strategy.py]: I would give a more specific name to this argument, like `archive_vcs_info`. What do you think?

File: lib/spack/spack/fetch_strategy.py (hunk 1 above)
Comment (matz-e) [reply lib/spack/spack/fetch_strategy.py]: While it could be used in a more generic way, `exclude` is specific to `VCSFetchStrategy.archive`. Looking around, it seems to be used only by subclasses, and a `git blame` suggests that intent, too. Thus I'd prefer to keep the code simpler…

**Comment by matz-e** [comment]
Done in https://github.com/spack/spack/commit/ad4309b7827004010c60ee5ee8332322a5899166

**Comment by matz-e** [comment]
Actually not done. I'll see about generalizing this once our deployment is running on v0.13

**Comment by alalazo** [comment]
Closing the PR as it seems stale. Feel free to reopen it or submit a new one in case you disagree.
